src/pip/_internal/req/req_set.py

```diff
@@ -110,9 +110,8 @@ def add_requirement(
                 )
 
         # This next bit is really a sanity check.
-        assert install_req.is_direct == (parent_req_name is None), (
-            "a direct req shouldn't have a parent and also, "
-            "a non direct req should have a parent"
+        assert not install_req.user_supplied or parent_req_name is None, (
+            "a direct req shouldn't have a parent"
```

Comment:
```suggestion
            "a user supplied req shouldn't have a parent"
```
Was it a deliberate choice to skip the "non direct should have a parent" side of the check here? If so, why?

Reply:
@pfmoore that is because constraints are also reqs and constraints are not considered to be user supplied reqs anymore (see also https://github.com/pypa/pip/pull/8026#issuecomment-612675047).

Reply:
Ah, I see. That makes sense.